### PR TITLE
Change to new MSD computation approach.

### PR DESCRIPTION
**Pull request by SamTov**
I have update all MSD computations to use a new, better performing TF approach. Changes include:

* Introduce general MSD method in the Calculator class
* Update all MSD based calculators to use new approach. Note, not all of them use the method in the calculator class as many require small changes. We can consider a better method for consolidation but for now it is still only one line of code.
* Fixed the set charge method in the experiment class and included type hints in the parameters to better facilitate use of the method in interactive sessions.

I have tested the implementation on Molten NaCl and KCl and the results match that of the previous method but are significantly faster. I have done this for self-diffusion and ionic conductivity. This covers all new approaches but I have not explicitly tested thermal conductivity and viscosity.

### CODE REVIEW

File: mdsuite/calculators/einstein_helfand_ionic_conductivity.py [hunk 1]
```diff
@@ -161,8 +164,7 @@ def _apply_operation(self, ensemble, index):
         -------
         MSD of the tensor_values.
         """
-        msd = (ensemble - (
-            tf.repeat(tf.expand_dims(ensemble[0], 0), self.data_range, axis=0))) ** 2
+        msd = tf.math.squared_difference(ensemble, ensemble[None, 0])
```
Comment (PythonFZ) [inline mdsuite/calculators/einstein_helfand_ionic_conductivity.py:167]: Why not use `Calculator._msd_operation`

File: mdsuite/calculators/einstein_helfand_thermal_conductivity.py [hunk 2]
```diff
@@ -158,8 +158,8 @@ def _apply_operation(self, ensemble, index):
         -------
         MSD of the tensor_values.
         """
-        msd = (ensemble - (
-            tf.repeat(tf.expand_dims(ensemble[0], 0), self.data_range, axis=0))) ** 2
+        msd = tf.math.squared_difference(ensemble, ensemble[None, 0])
```
Comment (PythonFZ) [inline mdsuite/calculators/einstein_helfand_thermal_conductivity.py:161]: Why not use `Calculator._msd_operation`

File: mdsuite/calculators/einstein_helfand_thermal_kinaci.py [hunk 3]
```diff
@@ -159,8 +159,8 @@ def _apply_operation(self, ensemble, index):
         -------
         MSD of the tensor_values.
         """
-        msd = (ensemble - (
-            tf.repeat(tf.expand_dims(ensemble[0], 0), self.data_range, axis=0))) ** 2
```
Comment (PythonFZ) [inline mdsuite/calculators/einstein_helfand_thermal_kinaci.py:163]: Why not use `Calculator._msd_operation`

**Review by PythonFZ — COMMENTED** [review]
I can't say much about the experiment changes, but the rest looks good.
Is there a reason you aren't using `Calculator._msd_operation` everywhere?

File: mdsuite/calculators/einstein_helfand_ionic_conductivity.py (hunk 1 above)
Comment (SamTov) [reply mdsuite/calculators/einstein_helfand_ionic_conductivity.py:167]: Yes, the dimension expansion is along differing axis so the option is either to have the computation in the calculator itself, or to have quite a large logic segment in the main _msd_operation method. We would need logic for square=True/False and then a check if system=True logic section under each. It isn't bad but it didn't seem elegant. I can think of a way to integrate it all.

Comment (PythonFZ) [reply mdsuite/calculators/einstein_helfand_ionic_conductivity.py:167]: Okay, in that case I don't see a reason, why it should be changed. Looks all good

**Review by PythonFZ — APPROVED** [review]

**Pull request merged**